File: src/conduit/loader/bigquery_loader.py
"""BigQuery destination writer."""

from __future__ import annotations

import logging
from pathlib import Path

import pyarrow as pa

from conduit.models import DestinationConfig

logger = logging.getLogger(__name__)


def load_bigquery(table: pa.Table, dest: DestinationConfig, base_dir: Path) -> None:
    """Write an Arrow table to a BigQuery table."""
    try:
        from google.cloud import bigquery
    except ImportError:
        raise ImportError(
            "google-cloud-bigquery is required for BigQuery destinations. "
            "Please ensure you are using the official Conduit ETL binary which bundles all drivers."
        )

    project = dest.config.get("project")
    dataset = dest.config.get("dataset")
    target_table = dest.config.get("table", dest.name)

    if not dataset:
        raise ValueError("BigQuery destination requires 'dataset' in config")

    table_ref = f"{dataset}.{target_table}"
    if project:
        table_ref = f"{project}.{table_ref}"

    client = bigquery.Client(project=project)

    write_disposition = (
        bigquery.WriteDisposition.WRITE_TRUNCATE
        if dest.mode == "full_refresh"
        else bigquery.WriteDisposition.WRITE_APPEND
    )

    job_config = bigquery.LoadJobConfig(
        write_disposition=write_disposition,
        source_format=bigquery.SourceFormat.PARQUET,
    )

    job = client.load_table_from_dataframe(
        table.to_pandas(),
        table_ref,
        job_config=job_config,
    )
    job.result()  # Wait for completion

    logger.info(
        "Loaded %d rows to BigQuery destination '%s' (%s)",
        len(table), dest.name, table_ref,
    )

File: src/ts/loader/bigquery_loader.ts
/**
 * BigQuery destination writer.
 */

// deno-lint-ignore-file no-explicit-any
import { BigQuery } from "@google-cloud/bigquery";
import type { DataTable, DestinationConfig } from "../models.ts";

export async function loadBigquery(
  table: DataTable,
  dest: DestinationConfig,
  _baseDir: string,
): Promise<void> {
  const project = dest.config.project as string | undefined;
  const dataset = dest.config.dataset as string;
  const targetTable = (dest.config.table as string) ?? dest.name;

  if (!dataset) {
    throw new Error("BigQuery destination requires 'dataset' in config");
  }

  const client = new BigQuery({ projectId: project });
  const tableRef = client.dataset(dataset).table(targetTable);

  if (dest.mode === "full_refresh") {
    try {
      await tableRef.delete();
    } catch {
      // Table might not exist
    }
  }

  // Insert rows in batches
  for (let i = 0; i < table.rows.length; i += dest.batch_size) {
    const batch = table.rows.slice(i, i + dest.batch_size);
    await tableRef.insert(batch as any[]);
  }

  const ref = project ? `${project}.${dataset}.${targetTable}` : `${dataset}.${targetTable}`;
  console.log(
    `Loaded ${table.rows.length} rows to BigQuery destination '${dest.name}' (${ref})`,
  );
}
